astropy/io/votable/tree.py

```diff
@@ -2567,18 +2586,32 @@ def is_empty(self):
         """
         return self._empty
 
-    def create_arrays(self, nrows=0, config=None):
+    def create_arrays(
+        self,
+        nrows=0,
+        config=None,
+        *,
+        colnumbers=None,
+    ):
         """
         Create a new array to hold the data based on the current set
         of fields, and store them in the *array* and member variable.
         Any data in the existing array will be lost.
 
         *nrows*, if provided, is the number of rows to allocate.
+        *colnumbers*, if provided, is the list of column indices to select.
+        By default, all columns are selected.
         """
         if nrows is None:
             nrows = 0
+        if colnumbers is None:
+            colnumbers = list(range(len(self.all_fields)))
 
-        fields = self.fields
+        for i, field in enumerate(self.all_fields):
+            if i in colnumbers and field not in self.fields:
+                self.fields.append(field)
```

Comment:
If this method is called repeatedly with different `colnumbers`, `self.fields` could end up out of order and with fields not in `colnumbers`.  Also, for future protection if we lock down public access to `self.fields`, the append should probably be on `self._fields` (e.g., if later we do something like replace `.append()` with `.add_field()`).

If we do anything to fix this now, it should probably be to totally rebuild `self.fields` from `colnumbers`, but I'm uncomfortable with these side effects.  Ideally creating arrays shouldn't modify any attributes of the table but also ideally the encapsulation (public vs. private) aspects of constructing a table could be cleaned up quite a bit, so I'm hesitant to do anything radical in this PR.

Reply:
good point. Reconstructing `self._fields` entirely seems safer and isn't really more complicated. How about this ?
```patch
diff --git a/astropy/io/votable/tree.py b/astropy/io/votable/tree.py
index 08dade28db..5df03cdf0e 100644
--- a/astropy/io/votable/tree.py
+++ b/astropy/io/votable/tree.py
@@ -2607,9 +2607,12 @@ class TableElement(
         if colnumbers is None:
             colnumbers = list(range(len(self.all_fields)))
 
-        for i, field in enumerate(self.all_fields):
-            if i in colnumbers and field not in self.fields:
-                self.fields.append(field)
+        new_fields = HomogeneousList(
+            Field,
+            values=[f for i, f in enumerate(self.all_fields) if i in colnumbers],
+        )
+        if new_fields != self.fields:
+            self._fields = new_fields
 
         fields = self.all_fields
```
We could also add a warning if reconstruction is triggered.

Reply:
That implementation looks reasonable to me.  It's interesting (and fine) that references to `all_fields` ensure that `all_fields` is caught up with anything in `fields` just prior to syncing in the other direction.  (I'm not a huge fan of properties since they make such side-effect non-obvious, but that's my own hangup.)

I can see the motivation for a warning in that it would be an usual and possibly unexpected circumstance, but I also wonder if we could come up with succinct wording that a user could understand and act on.  Unless the reconstruction happens due to a logic error we haven't noticed, the user would be doing some fairly unusual things to trigger this (as we'll see if you add a test case for this), so I think a warning is probably not worth the effort.

Reply:
I pushed it as is. Thank you so much for your thoughtful review !

Reply:
Thank you for doing the work and being so responsive to questions and comments!  Could you go ahead and create an issue for the proposed follow-on deprecation patch?

Reply:
It's getting late, but this will be top of my task list for tomorrow !

Reply:
No worries. I opened https://github.com/astropy/astropy/issues/16314 for you. Feel free to add details there tomorrow as you see fit. Thanks!